security/pkg/nodeagent/cache/secretcache.go

```diff
@@ -254,12 +257,23 @@ func (sc *SecretManagerClient) GenerateSecret(resourceName string) (secret *secu
 	}
 
 	if resourceName != RootCertReqResourceName {
-		// cache hit
 		if c := sc.cache.GetWorkload(); c != nil {
+			// cache hit
 			cacheLog.WithLabels("ttl", time.Until(c.ExpireTime)).Info("returned workload certificate from cache")
 			return c, nil
 		}
 		t0 := time.Now()
+		sc.generateMutex.Lock()
+		defer sc.generateMutex.Unlock()
+		if c := sc.cache.GetWorkload(); c != nil {
```

Comment:
Why the following lines are invoked twice (one from line 260 to 264, the other from line 268 to 272)? 
		if c := sc.cache.GetWorkload(); c != nil {
			cacheLog.WithLabels("ttl", time.Until(c.ExpireTime)).Info("returned workload certificate from cache")
			return c, nil
		}

Reply:
because the cache may be populated after we acquire the lock

Reply:
if there are two concurrent calls we want to wait for one to finish then load it's result from cache

Reply:
Thanks for the explanation!